morpheus/_lib/src/objects/table_info.cpp

```diff
@@ -34,137 +33,156 @@
-            m_index_names(std::move(index_names)),
-            m_column_names(std::move(column_names)) {}
-
-    const pybind11::object &TableInfo::get_parent_table() const  {
-        return m_parent->get_py_object();
-    }
-
-    const cudf::table_view &TableInfo::get_view() const  {
-        return m_table_view;
-    }
-
-    std::vector<std::string> TableInfo::get_index_names() const  {
-        return m_index_names;
-    }
-
-    std::vector<std::string> TableInfo::get_column_names() const  {
-        return m_column_names;
-    }
+/****** Component public implementations *******************/
+/****** TableInfo****************************************/
+TableInfo::TableInfo(std::shared_ptr<const IDataTable> parent,
+                     cudf::table_view view,
+                     std::vector<std::string> index_names,
+                     std::vector<std::string> column_names) :
+  m_parent(std::move(parent)),
+  m_table_view(std::move(view)),
+  m_index_names(std::move(index_names)),
+  m_column_names(std::move(column_names))
+{}
+
+const pybind11::object &TableInfo::get_parent_table() const
+{
+    return m_parent->get_py_object();
+}
+
+const cudf::table_view &TableInfo::get_view() const
+{
+    return m_table_view;
+}
+
+std::vector<std::string> TableInfo::get_index_names() const
+{
+    return m_index_names;
+}
+
+std::vector<std::string> TableInfo::get_column_names() const
+{
+    return m_column_names;
+}
+
+cudf::size_type TableInfo::num_indices() const
+{
+    return this->get_index_names().size();
+}
+
+cudf::size_type TableInfo::num_columns() const
+{
+    return this->get_column_names().size();
+}
+
+cudf::size_type TableInfo::num_rows() const
+{
+    return this->m_table_view.num_rows();
+}
+
+pybind11::object TableInfo::as_py_object() const
+{
+    const auto offset = m_table_view.column(0).offset();
+    const auto stop   = offset + this->num_rows();
 
-    cudf::size_type TableInfo::num_indices() const  {
-        return this->get_index_names().size();
-    }
+    {
+        namespace py = pybind11;
+        py::gil_scoped_acquire gil;
 
-    cudf::size_type TableInfo::num_columns() const  {
-        return this->get_column_names().size();
+        auto df          = this->get_parent_table();
+        auto index_slice = py::slice(py::int_(offset), py::int_(stop), py::none());
+        return df.attr("loc")[py::make_tuple(df.attr("index")[index_slice], m_column_names)];
     }
+}
 
-    cudf::size_type TableInfo::num_rows() const  {
-        return this->m_table_view.num_rows();
-    }
+void TableInfo::insert_columns(const std::vector<std::string> &column_names, const std::vector<TypeId> &column_types)
+{
+    CHECK(column_names.size() == column_types.size());
+    const auto num_existing_cols = m_column_names.size();
+    const auto num_rows          = m_table_view.num_rows();
 
-    pybind11::object TableInfo::as_py_object() const
+    // TODO figure out how to do this without the gil
```

Comment:
You could probably reproduce whatever cupy_zeros is doing in c++ without much problem (maybe?), kind of stuck since you have to work with the table object though... But, who knows, if "insert" is just attaching a blob of bytes to the object, then its likely we could avoid the gil for 99% of the process.

Reply:
I think in practice the output columns are already defined in the DF, so typically this code is avoided. So I think for this reason we decided it was OK to hold the gil until we can adopt a newer version of CUDF.

The change to this method I believe was just renaming "neo::TypeId" to "TypeId" but VSCode decided to apply some additional formatting to make the change look larger than it is.